### PR TITLE
Allow macro substitution in endpoint of disks  

### LINKED ISSUE #40951: Allow macro substitution in endpoint config parameter in storage_configuration section
It is convenient for managing huge clusters to allow using macros in endpoint config parameter in storage_configuration section similar to zookeeper paths - https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replication#creating-replicated-tables

This will allow to place files in S3 buckets on per shard basis for example.

### PR DESCRIPTION
**Pull request by SmitaRKulkarni**
Changelog category (leave one):
* Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow macro substitution in endpoint of disks resolve #40951

Implementation:
* Updated to substitute macros in endpoint for all disks (s3, hfs, web).
Testing:
* Added an integration test test/integration/test_endpoint_macro_substitution.

### CODE REVIEW

File: src/Coordination/KeeperSnapshotManagerS3.cpp [hunk 1]
```diff
@@ -64,7 +65,8 @@ void KeeperSnapshotManagerS3::updateS3Configuration(const Poco::Util::AbstractCo
 
         auto auth_settings = S3::AuthSettings::loadFromConfig(config_prefix, config);
 
-        auto endpoint = config.getString(config_prefix + ".endpoint");
+        String endpoint = config.getString(config_prefix + ".endpoint");
+        endpoint = macros->expand(endpoint);
```
Comment (alesapin) [inline src/Coordination/KeeperSnapshotManagerS3.cpp:69]: Why not do it in a single line? 
`macros->expand(config.getString(config_prefix + ".endpoint"))`

File: src/Coordination/KeeperDispatcher.cpp [hunk 2]
```diff
@@ -296,7 +296,7 @@ bool KeeperDispatcher::putRequest(const Coordination::ZooKeeperRequestPtr & requ
     return true;
 }
 
-void KeeperDispatcher::initialize(const Poco::Util::AbstractConfiguration & config, bool standalone_keeper, bool start_async)
+void KeeperDispatcher::initialize(const Poco::Util::AbstractConfiguration & config, bool standalone_keeper, bool start_async, const MultiVersion<Macros>::Version & macros)
```
Comment (alesapin) [inline src/Coordination/KeeperDispatcher.cpp:299]: We should add an explicit comment on what this macro affects.

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
LGTM

File: src/Coordination/KeeperSnapshotManagerS3.cpp (hunk 1 above)
Comment (SmitaRKulkarni) [reply src/Coordination/KeeperSnapshotManagerS3.cpp]: good point, addressed it

File: src/Coordination/KeeperDispatcher.cpp (hunk 2 above)
Comment (SmitaRKulkarni) [reply src/Coordination/KeeperDispatcher.cpp:299]: addressed it. I have added comments in header files where this new parameter is added

File: src/Disks/ObjectStorages/Web/registerDiskWebServer.cpp [hunk 3]
```diff
@@ -23,7 +26,8 @@ void registerDiskWebServer(DiskFactory & factory, bool global_skip_access_check)
         ContextPtr context,
         const DisksMap & /*map*/) -> DiskPtr
     {
-        String uri{config.getString(config_prefix + ".endpoint")};
+        String endpoint = context->getMacros()->expand(config.getString(config_prefix + ".endpoint"));
+        String uri{endpoint};
```
Comment (alesapin) [inline src/Disks/ObjectStorages/Web/registerDiskWebServer.cpp:30]:
```suggestion
        String uri = context->getMacros()->expand(config.getString(config_prefix + ".endpoint"));
```

File: src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp [hunk 4]
```diff
@@ -634,10 +637,11 @@ std::unique_ptr<IObjectStorage> S3ObjectStorage::cloneObjectStorage(
 {
     auto new_s3_settings = getSettings(config, config_prefix, context);
     auto new_client = getClient(config, config_prefix, context, *new_s3_settings);
+    String endpoint = context->getMacros()->expand(config.getString(config_prefix + ".endpoint"));
```
Comment (qoega) [inline src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:640]: This broke wildcards in s3 table function?

**Pull request merged**